projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts

```diff
@@ -0,0 +1,299 @@
+  return lajiMapObservationVisualization.obsCount.categories[idx].color;
+};
+
+const getRecordQualityColorIdx = (quality: string): number => {
+  switch (quality) {
+    case 'EXPERT_VERIFIED':
+      return 0;
+    case 'COMMUNITY_VERIFIED':
+      return 1;
+    case 'NEUTRAL':
+      return 2;
+    case 'UNCERTAIN':
+      return 3;
+    case 'ERRONEOUS':
+    default:
+      return 4;
+  }
+};
+
+const getRecordQualityColor = (qualities: string | string[]): string => {
+  if (!qualities) { return lajiMapObservationVisualization.recordQuality.categories[2].color; }
+  if (!Array.isArray(qualities)) { return lajiMapObservationVisualization.recordQuality.categories[getRecordQualityColorIdx(qualities)].color; }
+  const a = qualities.map(q => getRecordQualityColorIdx(q));
+  return lajiMapObservationVisualization.recordQuality.categories[Math.min(...a)].color;
+};
+
+const getRedlistColorIdx = (status: string): number => {
+  switch (status) {
+    case 'http://tun.fi/MX.iucnGroup1':
+      return 0;
+    case 'http://tun.fi/MX.iucnGroup2':
+      return 1;
+    case 'http://tun.fi/MX.iucnGroup3':
+      return 2;
+    case 'http://tun.fi/MX.iucnGroup4':
+      return 3;
+    case 'http://tun.fi/MX.iucnGroup5':
+    default:
+      return 4;
+  }
+};
+
+const getRedlistStatusColor = (statuses: string | string[]): string => {
+  if (!statuses) { return '#ffffff'; }
+  if (!Array.isArray(statuses)) { return lajiMapObservationVisualization.redlistStatus.categories[getRedlistColorIdx(statuses)].color; }
+  const a = statuses.map(s => getRedlistColorIdx(s));
+  return lajiMapObservationVisualization.redlistStatus.categories[Math.min(...a)].color;
+};
+
+const getIndividualCountColorIdx = (count: number): number => {
+  if (count === 0) {
+    return 0;
+  } else if (count <= 5) {
+    return 1;
+  } else if (count <= 20) {
+    return 2;
+  } else if (count <= 50) {
+    return 3;
+  } else if (count <= 100) {
+    return 4;
+  } else {
+    return 5;
+  }
+};
+
+const getIndividualCountColor = (count: number): string => (
+  lajiMapObservationVisualization.individualCount.categories[getIndividualCountColorIdx(count)].color
+);
+
+const getRecordAgeColorIdx = (year: number): number => {
+  for (let i = 0; i < yearsAgoBreakpoints.length; i++) {
+    const bp = yearsAgoBreakpoints[i];
+    if (year >= currentYear - bp) {
+      return i;
+    }
+  }
+  return yearsAgoBreakpoints.length - 1;
+};
+
+const sliceYear = (newestRecord: string): number => (
+  parseInt((newestRecord).slice(0, 4), 10)
+);
+
+const getRecordAgeColor = (newestRecords: string | string[]): string => {
+  if (!newestRecords) { return '#ffffff'; }
+  if (!Array.isArray(newestRecords)) { return lajiMapObservationVisualization.recordAge.categories[getRecordAgeColorIdx(sliceYear(newestRecords))].color; }
+  const a = newestRecords.map(s => sliceYear(s));
+  const year = Math.max(...a);
+  return lajiMapObservationVisualization.recordAge.categories[getRecordAgeColorIdx(year)].color;
```

Comment:
Looks familiar. Similarly to the other color getters I commented about, this could be simplified a bit (treat input as array, use default value). Now that I've read this far, I wonder if these functions could be refactored to use some shared logic. There are maybe enough differences that it's not worth it...